File: curriculum/13_api_gateway_circuit_breaker/node-impl/src/rateLimiter.ts
interface TokenBucket {
  tokens: number;
  lastRefill: number;
}

export class TenantLimiter {
  private buckets = new Map<string, TokenBucket>();

  constructor(private capacity: number, private refillPerSecond: number) {}

  allow(tenantId: string): boolean {
    const now = Date.now();
    let bucket = this.buckets.get(tenantId);
    if (!bucket) {
      bucket = { tokens: this.capacity, lastRefill: now };
      this.buckets.set(tenantId, bucket);
    }

    const elapsed = (now - bucket.lastRefill) / 1000;
    bucket.tokens = Math.min(this.capacity, bucket.tokens + elapsed * this.refillPerSecond);
    bucket.lastRefill = now;

    if (bucket.tokens >= 1) {
      bucket.tokens--;
      return true;
    }
    return false;
  }

  tokensRemaining(tenantId: string): number {
    const now = Date.now();
    const bucket = this.buckets.get(tenantId);
    if (!bucket) return this.capacity;

    const elapsed = (now - bucket.lastRefill) / 1000;
    return Math.min(this.capacity, bucket.tokens + elapsed * this.refillPerSecond);
  }

  resetAt(tenantId: string): Date {
    const tokens = this.tokensRemaining(tenantId);
    const needed = this.capacity - tokens;
    const seconds = needed / this.refillPerSecond;
    return new Date(Date.now() + seconds * 1000);
  }
}

File: curriculum/13_api_gateway_circuit_breaker/go-impl/gateway/ratelimiter.go
package gateway

import (
	"sync"
	"time"
)

type TenantLimiter struct {
	mu       sync.RWMutex
	capacity int
	refill   float64
	buckets  map[string]*tokenBucket
}

type tokenBucket struct {
	tokens     float64
	lastRefill time.Time
}

func NewTenantLimiter(capacity int, refillPerSecond float64) *TenantLimiter {
	return &TenantLimiter{
		capacity: capacity,
		refill:   refillPerSecond,
		buckets:  make(map[string]*tokenBucket),
	}
}

func (tl *TenantLimiter) Allow(tenantID string) bool {
	tl.mu.Lock()
	defer tl.mu.Unlock()

	b, ok := tl.buckets[tenantID]
	if !ok {
		b = &tokenBucket{tokens: float64(tl.capacity), lastRefill: time.Now()}
		tl.buckets[tenantID] = b
	}

	now := time.Now()
	elapsed := now.Sub(b.lastRefill).Seconds()
	b.tokens = min(float64(tl.capacity), b.tokens+elapsed*tl.refill)
	b.lastRefill = now

	if b.tokens >= 1 {
		b.tokens--
		return true
	}
	return false
}

func (tl *TenantLimiter) TokensRemaining(tenantID string) float64 {
	tl.mu.RLock()
	defer tl.mu.RUnlock()

	b, ok := tl.buckets[tenantID]
	if !ok {
		return float64(tl.capacity)
	}

	now := time.Now()
	elapsed := now.Sub(b.lastRefill).Seconds()
	tokens := min(float64(tl.capacity), b.tokens+elapsed*tl.refill)
	return tokens
}

func (tl *TenantLimiter) ResetAt(tenantID string) time.Time {
	tl.mu.RLock()
	defer tl.mu.RUnlock()

	b, ok := tl.buckets[tenantID]
	if !ok {
		return time.Now()
	}

	now := time.Now()
	elapsed := now.Sub(b.lastRefill).Seconds()
	tokens := min(float64(tl.capacity), b.tokens+elapsed*tl.refill)
	needed := float64(tl.capacity) - tokens
	seconds := needed / tl.refill
	return now.Add(time.Duration(seconds) * time.Second)
}